File: UnionPressOnSharp/UnionPressOnSharp/Forms/Presenters/TypePresenter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using UnionPressOnSharp.Forms.Models;
using UnionPressOnSharp.Forms.Views;
using UnionPressOnSharp.Forms.Presenters.Common;

namespace UnionPressOnSharp.Forms.Presenters
{
    public class TypePresenter
    {
        private ITypeView typeView;
        private ITypeRepository  typeRepository;
        private BindingSource typeBindingSource;
        private IEnumerable<TypeModel>typeList;

        public TypePresenter(ITypeView typeView, ITypeRepository typeRepository)
        {
            this.typeBindingSource = new BindingSource();
            this.typeView = typeView;
            this.typeRepository = typeRepository;

            this.typeView.SearchDataEvent += Search;
            this.typeView.AddNewDataEvent += AddNew;
            this.typeView.EditDataEvent += LoadSelectedToEdit;
            this.typeView.DeleteDataEvent += DeleteSelected;
            this.typeView.SaveDataEvent += Save;
            this.typeView.CancelEvent += CancelAction;
            this.typeView.SetTypeListBindingSource(typeBindingSource);

            LoadAllList();


        }


        private void LoadAllList()
        {
            typeList = typeRepository.GetAllType();
            typeBindingSource.DataSource = typeList;
        }
        private void Search(object sender, EventArgs e)
        {
            bool emptyValue = string.IsNullOrWhiteSpace(this.typeView.SearchValue);
            if (emptyValue == false)
                typeList = typeRepository.GetByValue(this.typeView.SearchValue);
            else typeList = typeRepository.GetAllType();
            typeBindingSource.DataSource = typeList;
        }
        private void CancelAction(object sender, EventArgs e)
        {
            typeView.isEditData = false;
            CleanviewFields();
        }
        private void Save(object sender, EventArgs e)
        {
            var typeModel = new TypeModel();
 
            typeModel.Type = typeView.TypePresss;
            try
            {
                new Common.ModelDataValidation().Validate(typeModel);
                if (typeView.isEditData)//Edit model
                {
                    typeRepository.EditData(typeModel);
                    typeView.Message = "Тип отредактирован";
                }
                else //Add new model
                {
                    typeRepository.AddData(typeModel);
                    typeView.Message = "Тип добавлен";
                }
                typeView.isSuccessful = true;
                LoadAllList();
                CleanviewFields();
            }
            catch (Exception ex)
            {
                typeView.isSuccessful = false;
                typeView.Message = ex.Message;
            }
        }

        private void CleanviewFields()
        {
            typeView.TypePresss = "";
        }
        private void DeleteSelected(object sender, EventArgs e)
        {
            try
            {
                var type = (TypeModel)typeBindingSource.Current;
                typeRepository.DeleteData(type.Type);
                typeView.isSuccessful = true;
                typeView.Message = "Тип удален";
                LoadAllList();
            }
            catch (Exception)
            {
                typeView.isSuccessful = false;
                typeView.Message = "Ошибка, не удалось удалить тип";
            }
        }
        private void LoadSelectedToEdit(object sender, EventArgs e)
        {
            var type = (TypeModel)typeBindingSource.Current;

            typeView.TypePresss = type.Type;;
            typeView.isEditData = true;
        }
        private void AddNew(object sender, EventArgs e)
        {
            typeView.isEditData = false;
        }


    }
}

File: UnionPressOnC/UnionPressOnC/Forms/Presenters/TypePresenter.cpp
#include <string>
#include "../Views/ITypeView.h"
#include "../Models/ITypeRepository.h"
using namespace UnionPressOnC;
using namespace  System;
using namespace  System::Collections::Generic;
using namespace  System::Windows::Forms;

namespace UnionPressOnC
{
    public class TypePresenter
    {
    private:
         ITypeView typeView;
         ITypeRepository  typeRepository;
         BindingSource typeBindingSource;
         IEnumerable<TypeModel>typeList;

    public:
        
        TypePresenter(ITypeView typeView, ITypeRepository typeRepository)
        {
            this->typeBindingSource = gcnew BindingSource();
            this->typeView = typeView;
            this->typeRepository = typeRepository;

            this->typeView.SearchDataEvent += Search;
            this->typeView.AddNewDataEvent += AddNew;
            this->typeView.EditDataEvent += LoadSelectedToEdit;
            this->typeView.DeleteDataEvent += DeleteSelected;
            this->typeView.SaveDataEvent += Save;
            this->typeView.CancelEvent += CancelAction;
            this->typeView.SetTypeListBindingSource(typeBindingSource);

            LoadAllList();
        }


    private:
            
        void LoadAllList()
        {
            typeList = typeRepository.GetAllType();
            typeBindingSource.DataSource = typeList;
        }
        
        void Search(Object sender, EventArgs e)
        {
            bool emptyValue = String::IsNullOrWhiteSpace(this->typeView.getSearchValue());
            if (emptyValue == false)
                typeList = typeRepository.GetByValue(this->typeView.getSearchValue());
            else typeList = typeRepository.GetAllType();
            typeBindingSource.DataSource = typeList;
        }
        
        void CancelAction(Object sender, EventArgs e)
        {
            typeView.setIsEditData(false);
            CleanviewFields();
        }
        
        void Save(Object sender, EventArgs e)
        {
            TypeModel typeModel;

            typeModel.type = typeView.getTypePresss();
            try
            {
                //new Common.ModelDataValidation().Validate(typeModel);
                if (typeView.getIsEditData())//Edit model
                {
                    typeRepository.EditData(typeModel);
                    typeView.setMessage("Тип отредактирован");
                }
                else //Add new model
                {
                    typeRepository.AddData(typeModel);
                    typeView.setMessage("Тип добавлен");
                }
                typeView.setIsSuccessful(true);
                LoadAllList();
                CleanviewFields();
            }
            catch (Exception^ ex)
            {
                typeView.setIsSuccessful(false);
                typeView.setMessage(ex->Message);
            }
        }

        
        void CleanviewFields()
        {
            typeView.setTypePresss("");
        }
        
        void DeleteSelected(Object sender, EventArgs e)
        {
            try
            {
                TypeModel type = (TypeModel)typeBindingSource.Current;
                typeRepository.DeleteData(type.type);
                typeView.setIsSuccessful(true);
                typeView.setMessage("Тип удален");
                LoadAllList();
            }
            catch (Exception^)
            {
                typeView.setIsSuccessful(false);
                typeView.setMessage("Ошибка, не удалось удалить тип");
            }
        }
        
        void LoadSelectedToEdit(Object sender, EventArgs e)
        {
            TypeModel type = (TypeModel)typeBindingSource.Current;

            typeView.setTypePresss(type.type);
            typeView.setIsEditData(true);
        }
        
        void AddNew(Object sender, EventArgs e)
        {
            typeView.setIsEditData(false);
        }


    }
}
